### PR TITLE
Make Projector batch aware

### PR DESCRIPTION
**Pull request by dxoigmn**
# What does this PR do?

The current `Projector` is not batch/tuple aware, which is necessary for attacks on datasets where input images have different shapes (e.g., COCO). This PR makes it so.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] `make test`

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [x] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by mzweilin** [comment]
Do you have a plan for modality-aware?

**Comment by dxoigmn** [comment]
Nope.

File: mart/attack/projector.py
Comment (mzweilin) [inline mart/attack/projector.py:45]: Do you think it's better to make it a dictionary so that it's more easily configurable in Hydra? Like that in Enforcer: https://github.com/IntelLabs/MART/blob/70e62d901f2a27327dca517ed21a12d81dd49040/mart/attack/enforcer.py#L100

The keys in the Enforcer constraints dictionary are not used. Maybe we can use that for modality configuration?

Comment (dxoigmn) [reply mart/attack/projector.py:45]: I think that should live in another PR.

Comment (dxoigmn) [reply mart/attack/projector.py:45]: My read of the Projectors class is that they need a wholesale redesign because there's a lot of mixing of functionality instead of focusing on composition.

**Comment by mzweilin** [comment]
I have made a shared modality-batch-aware mechanism in #115 for Initializer, Projector and Composer.

**Comment by dxoigmn** [comment]
> I have made a shared modality-batch-aware mechanism in #115 for Initializer, Projector and Composer.

Okay. Is the idea that this PR should be updated to use that? I also saw there was an opportunity to DRY the logic but wanted to wait until after we get everything merged. I'll try to read through what you wrote though.

**Comment by dxoigmn** [comment]
@mzweilin: I think it's better to merge this now than wait for the generic dispatch to be done since it updates the tests to conform to the existing interface.

**Review by mzweilin — APPROVED** [review]
LGTM

**Pull request merged**
